### PR TITLE
Consume CLI flags when parsing them.

### PR DESCRIPTION
**Pull request by jettisonjoe**
Proposed solution to enable using 3rd party libraries that blow up when encountering unknown command line arguments.

### CODE REVIEW

**Comment by fahhem** [comment]
Is this to deal with gflags? The way I'd recommend with that it is to pass the arguments you want to `gflags.Flags()`. Instead of `Flags(sys.argv)`, pass the `unknown` variable you get from `parse_known_args()`

**Comment by grybmadsci** [comment]
I like the idea, but it may be better to make this an optional thing rather than the default.  for example, the way this broke the CI unit test is fairly likely to happen, i think...

**Comment by jettisonjoe** [comment]
@fahhem, this is to deal with the general case of coexisting with other frameworks and libraries that also take command-line args. In my case it's Mobly. We have a parallel initiative in progress to patch Mobly so it doesn't blow up on unrecognized flags, but this would enable OpenHTF to hide its flags from other badly-behaving consumers of command line args.

@grybmadsci, Fair point, I'll work up an optional version and see how that looks.

**Comment by jettisonjoe** [comment]
Updated to make consuming happen conditionally on an argument to the consuming parser's constructor, but had to do additional work on the unit test to make that pass. Turns out the sys.argv mods that test was doing weren't working right, as other tests getting imported first would also import the conf module, and so the conf import after the sys.argv mod in conf_test.py was a no-op. PTAL.

File: openhtf/__init__.py [hunk 1]
```diff
@@ -328,9 +328,10 @@ def create_arg_parser(add_help=False):
           'My args title', parents=[openhtf.create_arg_parser()])
   >>> parser.parse_args()
   """
-  return argparse.ArgumentParser('OpenHTF-based testing', parents=[
+  return util.argv.SysArgvConsumingParser('OpenHTF-based testing', parents=[
       conf.ARG_PARSER, phase_executor.ARG_PARSER, logs.ARG_PARSER],
-      add_help=add_help)
+      add_help=add_help,
+      consume=True)
```
Comment (grybmadsci) [inline openhtf/__init__.py]: maybe take consume to create_arg_parser() and default it to True there?  this still isn't especially optional if the optionality isn't propagated to the external API.

**Comment by grybmadsci** [comment]
just the one comment, otherwise LGTM

**Comment by jettisonjoe** [comment]
Broke it out further to create_arg_parser. Would maybe be nice to have it _actually_ externally settable via `configure()` but there's not a great way to do that since it's necessarily not part of the TestOptions interface. We could make it an additional magic kwarg that `configure()` takes, or just leave it where I have it for now.

**Comment by jettisonjoe** [comment]
PTAL

File: openhtf/__init__.py [hunk 2]
```diff
@@ -319,7 +319,7 @@ def plug_types(self):
     return {plug.cls for phase in self.phases for plug in phase.plugs}
 
 
-def create_arg_parser(add_help=False):
+def create_arg_parser(add_help=False, consume=True):
```
Comment (fahhem) [inline openhtf/__init__.py]: I thought we agreed it wouldn't be the default?

File: openhtf/util/argv.py [hunk 3]
```diff
@@ -17,10 +17,39 @@
...
+  """ArgumentParser variant that can consume known arguments from sys.argv.
+
+  Args (in addition to the normal argparse.ArgumentParser args):
+    consume: If True, consumes known args from sys.argv upon parsing.
+  """
+  def __init__(self, *args, **kwargs):
+    self._consume = kwargs.pop('consume')
+    super(SysArgvConsumingParser, self).__init__(*args, **kwargs)
+
+  def parse_args(self):
```
Comment (fahhem) [inline openhtf/util/argv.py]: why are you overriding this? Also, it doesn't return anything, so you're just breaking the API

File: openhtf/__init__.py [hunk 4]
```diff
@@ -328,9 +328,10 @@ def create_arg_parser(add_help=False):
           'My args title', parents=[openhtf.create_arg_parser()])
   >>> parser.parse_args()
   """
-  return argparse.ArgumentParser('OpenHTF-based testing', parents=[
+  return util.argv.SysArgvConsumingParser('OpenHTF-based testing', parents=[
```
Comment (fahhem) [inline openhtf/__init__.py]: why not choose the class based on whether consume is true?

File: setup.py [hunk 5]
```diff
@@ -160,7 +160,7 @@ def run_tests(self):
 
 setup(
     name='openhtf',
-    version='1.1.0',
+    version='1.1.1',
```
Comment (fahhem) [inline setup.py]: are we supposed to update this in every commit?

**Comment by jettisonjoe** [comment]
The offending other library accepted our fix to their arg parsing code, so this is no longer an issue. Retiring this PR for now.

**Pull request closed without merging**